### PR TITLE
Support MessageBodyWriter/Reader in the HTTP client

### PR DESCRIPTION
**Pull request by yawkat**
This implements the reader/writer API for the client. There's also some server changes that I missed.

Some changes from MediaTypeCodec to MessageBodyHandler are still missing, such as for form upload conversions.

### CODE REVIEW

File: http-client/src/test/groovy/io/micronaut/http/client/ServerRedirectSpec.groovy [hunk 1]
```diff
@@ -326,7 +332,7 @@ class ServerRedirectSpec extends Specification {
         @Produces([MediaType.TEXT_EVENT_STREAM, MediaType.APPLICATION_JSON_STREAM])
         HttpResponse<Flux<?>> title(final HttpHeaders headers) {
             if (headers.accept().contains(MediaType.TEXT_EVENT_STREAM_TYPE)) {
-                return HttpResponse.ok(Flux.just("The Stand")).contentType(MediaType.TEXT_EVENT_STREAM)
+                return HttpResponse.ok(Flux.just(Event.of("The Stand"))).contentType(MediaType.TEXT_EVENT_STREAM)
```
Comment (dstepanov) [inline http-client/src/test/groovy/io/micronaut/http/client/ServerRedirectSpec.groovy:335]: Is it required now?

File: http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy [hunk 2]
```diff
@@ -55,7 +55,7 @@ class MutateRequestClientFilterSpec extends Specification {
 
     void "test mutate stream request URI"() {
         expect:
-        myClient.stream().blockFirst() == "xxxxxxxxxxx"
+        myClient.stream().collectList().block().join("") == '["xxxxxxxxxxx"]'
```
Comment (dstepanov) [inline http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy:58]: Why is this change?

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java [hunk 3]
```diff
@@ -307,30 +298,20 @@ private void encodeHttpResponse(
...
-                NettyBodyWriter<Object> closure = wrap(messageBodyWriter);
-                handleMissingConnectionHeader(response, nettyRequest, outboundAccess);
-                if (closure.isBlocking()) {
-                    MediaType finalResponseMediaType = responseMediaType;
-                    getIoExecutor().execute(() -> writeNettyMessageBody(nettyRequest.getChannelHandlerContext(), nettyRequest, (MutableHttpResponse<Object>) response, responseBodyType, finalResponseMediaType, body, closure, outboundAccess));
-                } else {
-                    writeNettyMessageBody(nettyRequest.getChannelHandlerContext(), nettyRequest, (MutableHttpResponse<Object>) response, responseBodyType, responseMediaType, body, closure, outboundAccess);
-                }
+            Argument<Object> actualResponseType;
+            if (messageBodyWriter == null || !responseBodyType.isInstance(body) || !messageBodyWriter.isWriteable(responseBodyType, responseMediaType)) {
```
Comment (dstepanov) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:302]: Maybe `messageBodyWriter` can be reused in `RouteInfo`?

File: http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java [hunk 4]
```diff
@@ -17,205 +17,101 @@
...
-                        writer = beanLocator.getBean(exactMatch.iterator().next());
+    protected <T> MessageBodyWriter<T> findWriterImpl(Argument<T> type, List<MediaType> mediaTypes) {
+        Collection<BeanDefinition<MessageBodyWriter>> beanDefinitions = beanLocator.getBeanDefinitions(Argument.of(MessageBodyWriter.class, type), new MediaTypeQualifier<>(type, mediaTypes, Produces.class));
+        if (beanDefinitions.size() == 1) {
+            return beanLocator.getBean(beanDefinitions.iterator().next());
+        } else {
+            List<BeanDefinition<MessageBodyWriter>> exactMatch = beanDefinitions.stream()
+                .filter(d -> {
+                    List<Argument<?>> typeArguments = d.getTypeArguments(MessageBodyWriter.class);
+                    if (typeArguments.isEmpty()) {
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java:100]: @graemerocher Do we have some qualifier from `Qualifiers.` that can do this match? I think the correct way would be `Qualifiers.byQualifiers(new MediaTypeQualifier(...), byTypeArgumentMatching(type))`

File: http/src/main/java/io/micronaut/http/body/DynamicWriter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public MessageBodyWriter<Object> createSpecific(Argument<Object> type) {
+        return registry.findWriter(type, mediaTypes).orElse(this);
+    }
+
+    public MessageBodyWriter<Object> find(Argument<Object> type, MediaType mediaType, Object object) {
+        Optional<MessageBodyWriter<Object>> specific = registry.findWriter(type, List.of(mediaType));
+        if (specific.isPresent() && !(specific.get() instanceof DynamicWriter)) {
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/body/DynamicWriter.java]: If the writer is shared maybe we can store `lastFound` and `lastArgument` if it matches, return it, avoiding the writer.

File: http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    }
+
+    protected abstract <T> MessageBodyReader<T> findReaderImpl(Argument<T> type, List<MediaType> mediaTypes);
+
+    @SuppressWarnings({"unchecked"})
+    @Override
+    public <T> Optional<MessageBodyReader<T>> findReader(Argument<T> type, List<MediaType> mediaTypes) {
+        HandlerKey<T> key = new HandlerKey<>(type, mediaTypes);
+        MessageBodyReader<?> messageBodyReader = readers.get(key);
+        if (messageBodyReader == null) {
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java:92]: Would it make sense to have two maps:
- 1. All readers/writers contributed at the startup - simple hashmap - fast, the most common scenario
- 2. Precalculated readers/writers - requires a concurrent map

The minus is that we would need to access two maps in the worse scenario.
If the readers/writers are cached in the route info, this is unnecessary.

File: http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            HandlerKey<?> that = (HandlerKey<?>) o;
+            return type.equalsType(that.type) && mediaTypes.equals(that.mediaTypes);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(type.typeHashCode(), mediaTypes);
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java]: You can use `io.micronaut.core.util.ObjectUtils` that avoids new array allocation. You might want to store the hash code.

File: http/src/main/java/io/micronaut/http/body/TextPlainHandler.java [hunk 8]
```diff
@@ -44,26 +44,26 @@
 @Consumes(MediaType.TEXT_PLAIN)
 @Singleton
 @Experimental
-public final class TextPlainHandler implements MessageBodyHandler<String> {
+public final class TextPlainHandler implements MessageBodyHandler<CharSequence> {
```
Comment (dstepanov) [inline http/src/main/java/io/micronaut/http/body/TextPlainHandler.java:47]: We might want to have an additional `MessageBodyHandler<String>` so the handler is found as a direct match.

**Review by dstepanov — CHANGES_REQUESTED** [review]

**Comment by dstepanov** [comment]
Can you please add comments in all modified tests if it's a breaking change or a test improvement?

File: http-client/src/test/groovy/io/micronaut/http/client/ServerRedirectSpec.groovy (hunk 1 above)
Comment (yawkat) [reply http-client/src/test/groovy/io/micronaut/http/client/ServerRedirectSpec.groovy:335]: String is a raw type, the new implementation thinks it's already serialized. It's more consistent with other implementations (e.g. json) so I think it's fine.

File: http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy (hunk 2 above)
Comment (yawkat) [reply http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy:58]: bc String is a raw type, it doesn't deserialize it from array anymore. the controller already added the quotes manually so the test case was a bit weird in the first place

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java (hunk 3 above)
Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:302]: it should be in most cases. dynamic writer is used only as fallback, eg when theres an error where the route writer does not match the type of the actual response body.

File: http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java (hunk 6 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java:92]: the problem is that the matching is kind of fuzzy. for example the json mapper matches for any body type, and is cached separately for each type. so adding it for the base type (Object) is basically pointless.

The map is kind of like the cache in ConversionService.

File: http-client/src/test/groovy/io/micronaut/http/client/HttpGetSpec.groovy [hunk 9]
```diff
@@ -311,7 +311,7 @@ class HttpGetSpec extends Specification {
         String body = res.getBody(String).orElse(null)
 
         then:
-        body == null
```
Comment (yawkat) [inline http-client/src/test/groovy/io/micronaut/http/client/HttpGetSpec.groovy:314]: behavior change, but not a breaking change. body is now available forever.

Comment (yawkat) [reply http-client/src/test/groovy/io/micronaut/http/client/HttpGetSpec.groovy:314]: i gave up on lifecycle management, but it's not worse than before, because the body was stored as byte[] before anyway :)

File: http-client/src/test/groovy/io/micronaut/http/client/JsonStreamSpec.groovy [hunk 10]
```diff
@@ -209,7 +209,6 @@ class JsonStreamSpec  extends Specification {
         Publisher<Book> list();
 
         @Post(uri = "/count", processes = MediaType.APPLICATION_JSON_STREAM)
-        @SingleResult
         Publisher<LibraryStats> count(@Body Flux<Book> theBooks)
```
Comment (yawkat) [inline http-client/src/test/groovy/io/micronaut/http/client/JsonStreamSpec.groovy:212]: breaking change, APPLICATION_JSON_STREAM doesn't make sense for parsing to single result. it gives an error now.

File: http-client/src/test/groovy/io/micronaut/http/client/StreamRequestSpec.groovy [hunk 11]
```diff
@@ -273,7 +274,7 @@ class StreamRequestSpec extends Specification {
 
         @Post("/pojo-flowable-error")
         Publisher<Book> pojoReactiveSequenceError(@Header MediaType contentType, @Body Publisher<Book> books) {
-            return Flux.from(books).flatMap({ Book book ->
```
Comment (yawkat) [inline http-client/src/test/groovy/io/micronaut/http/client/StreamRequestSpec.groovy:276]: no change, concatMap is better but it makes no difference here

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/context/ServerRequestContextSpec.groovy [hunk 12]
```diff
@@ -109,30 +113,30 @@ class ServerRequestContextSpec extends Specification {
         @Get("/method")
         String method() {
             def request = ServerRequestContext.currentRequest().orElseThrow { -> new RuntimeException("no request") }
-            request.uri
```
Comment (yawkat) [inline http-server-netty/src/test/groovy/io/micronaut/http/server/netty/context/ServerRequestContextSpec.groovy:112]: technically a breaking change, but this was a test bug. the methods declare return type String but actually returned URI.

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/threading/ThreadSelectionSpec.groovy [hunk 13]
```diff
@@ -163,7 +163,7 @@ class ThreadSelectionSpec extends Specification {
         String alterflowable()
 
         @Post(uri = "/alterflowablePost", produces = MediaType.TEXT_PLAIN)
-        String alterflowablePost(String body)
```
Comment (yawkat) [inline http-server-netty/src/test/groovy/io/micronaut/http/server/netty/threading/ThreadSelectionSpec.groovy:166]: breaking change

File: http-server-tck/src/main/java/io/micronaut/http/server/tck/tests/ResponseStatusTest.java [hunk 14]
```diff
@@ -48,7 +48,7 @@ public class ResponseStatusTest {
     @Test
     void testConstraintViolationCauses400() throws IOException {
         asserts(SPEC_NAME,
-            HttpRequest.POST("/response-status/constraint-violation", Collections.emptyMap()).header(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_PLAIN),
```
Comment (yawkat) [inline http-server-tck/src/main/java/io/micronaut/http/server/tck/tests/ResponseStatusTest.java:51]: breaking change. the text/plain codec cannot serialize a map. previously it would fall back to json i think, now it will fail. the body is unused in the test anyway.

File: test-suite-groovy/src/test/groovy/io/micronaut/docs/server/consumes/ConsumesControllerSpec.groovy [hunk 15]
```diff
@@ -76,7 +77,7 @@ class ConsumesControllerSpec extends Specification {
                         .contentType(MediaType.TEXT_PLAIN))
 
         then:
-        noExceptionThrown()
```
Comment (yawkat) [inline test-suite-groovy/src/test/groovy/io/micronaut/docs/server/consumes/ConsumesControllerSpec.groovy:79]: breaking change, can't write arbitrary type (Book) as text/plain anymore

File: http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java (hunk 4 above)
Comment (graemerocher) [reply http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java:100]: yes you can have a composite qualifier, this could work

File: test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy [hunk 16]
```diff
@@ -102,7 +102,7 @@ class ProxyRequestSpec extends Specification {
         response.body().text == 'Hello John XXX'
 
         when:"A POST request is sent with an invalid content type"
-        response = client.exchange(HttpRequest.POST("/proxy/post/json", new Message(text: "John")).contentType(MediaType.TEXT_PLAIN), Message).blockFirst()
```
Comment (yawkat) [inline test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy:105]: breaking change, can't write arbitrary type (Book) as text/plain anymore

File: test-suite/src/test/java/io/micronaut/docs/function/client/aws/LocalFunctionInvokeSpec.java [hunk 17]
```diff
@@ -80,10 +79,11 @@ interface MathClient {
 
         //tag::functionRnd[]
         @Named("round")
-        int rnd(float value);
```
Comment (yawkat) [inline test-suite/src/test/java/io/micronaut/docs/function/client/aws/LocalFunctionInvokeSpec.java:83]: this change used to be necessary, don't think it is anymore. float and other primitive types can be written to text/plain still, for compat reasons.

File: test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy (hunk 16 above)
Comment (graemerocher) [reply test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy:105]: can you explain why?

File: http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy (hunk 2 above)
Comment (yawkat) [reply http-client/src/test/groovy/io/micronaut/http/client/filter/MutateRequestClientFilterSpec.groovy:58]: to explain more:

The controller returns a Flux<String>, composed of a single string "xxx" including the quotes. The http stack does not serialize the string as json because it's a raw type (serializing it would yield double quoting "\"xxx\""). However it does add the brackets because it's not byte[] or ByteBuffer (yes it's weird, but this is the most compatible behavior). This hasn't changed.

On the client side, it used to split up the input and parse the string fully as json. This has changed, it treats it as a raw type because it's String, so it just yields the unparsed body now.

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java [hunk 18]
```diff
@@ -498,12 +510,21 @@ public MediaTypeCodecRegistry getMediaTypeCodecRegistry() {
...
     }
 
+    public final MessageBodyHandlerRegistry getHandlerRegistry() {
+        return handlerRegistry;
+    }
+
+    public final void setHandlerRegistry(MessageBodyHandlerRegistry handlerRegistry) {
+        this.handlerRegistry = handlerRegistry;
+    }
+
```
Comment (graemerocher) [inline http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]: javadoc / nullability annotations

File: http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java [hunk 19]
```diff
@@ -156,65 +155,23 @@ public <T> Optional<T> getBody(Class<T> type) {
         return getBody(Argument.of(type));
     }
 
-    /**
-     * @return The Netty native response object
-     */
-    public FullHttpResponse getNativeResponse() {
-        return nettyHttpResponse;
-    }
```
Comment (graemerocher) [inline http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java:164]: Why this change?

File: http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java [hunk 20]
```diff
@@ -264,20 +217,9 @@ private <T> Optional convertByteBuf(ByteBuf content, Argument<T> type) {
...
-                    return Optional.of(bodyBytes);
-                } else {
-                    Optional<MediaTypeCodec> foundCodec = mediaTypeCodecRegistry.findCodec(contentType.get());
-                    if (foundCodec.isPresent()) {
-                        MediaTypeCodec codec = foundCodec.get();
-                        return Optional.of(codec.decode(type, bodyBytes));
-                    }
-                }
+            Optional<MessageBodyReader<T>> reader = handlerRegistry.findReader(type, List.of(contentType.get()));
+            if (reader.isPresent() && reader.get().isReadable(type, contentType.get())) {
```
Comment (graemerocher) [inline http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java]: split into 2 `if` blocks to avoid the multiple calls to `get` also a extract a variable for `contentType` to avoid multiple calls

File: http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawHandler.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * @since 4.0.0
+ * @author Jonas Konrad
+ */
+@Internal
+@Singleton
+@Experimental
+@BootstrapContextCompatible
+@Bean(typed = RawMessageBodyHandler.class)
+public class ByteBufRawHandler implements RawMessageBodyHandler<ByteBuf> {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawHandler.java]: can be `final`. Rename to `ByteBufRawMessageBodyHandler`

File: http-netty/src/main/java/io/micronaut/http/netty/body/NettyJsonStreamHandler.java [hunk 22]
```diff
@@ -33,23 +36,29 @@
...
 import java.io.InputStream;
+import java.io.OutputStream;
 import java.util.List;
 
 @Internal
 @Singleton
 @Produces(MediaType.APPLICATION_JSON_STREAM)
 @Consumes(MediaType.APPLICATION_JSON_STREAM)
-class NettyJsonStreamHandler<T> implements MessageBodyReader<T>, ChunkedMessageBodyReader<T> {
+public class NettyJsonStreamHandler<T> implements MessageBodyHandler<T>, ChunkedMessageBodyReader<T>, CustomizableNettyJsonHandler {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/body/NettyJsonStreamHandler.java]: can be `final`?

File: http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+ */
+package io.micronaut.http.body;
+
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.io.buffer.ByteBufferFactory;
+import io.micronaut.core.type.Argument;
+import io.micronaut.http.MediaType;
+import io.micronaut.runtime.ApplicationConfiguration;
+import org.jetbrains.annotations.Nullable;
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java]: use Micronaut nullable

File: http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * {@link MessageBodyHandlerRegistry} implementation that does not need an application context.
+ *
+ * @author Jonas Konrad
+ * @since 4.0.0
+ */
+@Internal
+@Experimental
+public final class ContextlessMessageBodyHandlerRegistry extends RawMessageBodyHandlerRegistry {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java:39]: Since this is only used by the client can it be moved to the client module and made package private?

File: http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+        super(Stream.concat(Stream.of(new RawStringHandler(applicationConfiguration), new RawByteArrayHandler(), new RawByteBufferHandler(byteBufferFactory)), Stream.of(otherRawHandlers)).toList());
+    }
+
+    /**
+     * Add a {@link MessageBodyHandler} for the given media type.
+     *
+     * @param mediaType The media type the handler applies to
+     * @param handler   The handler
+     */
+    public void add(MediaType mediaType, MessageBodyHandler<?> handler) {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java]: nullability annotations for the arguments

File: http/src/main/java/io/micronaut/http/body/DynamicWriter.java [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * Placeholder {@link MessageBodyWriter} implementation that decides which writer to use based on
+ * the dynamic (runtime) type of the body. Used as fallback where the type is not known statically.
+ *
+ * @author Jonas Konrad
+ * @since 4.0.0
+ */
+@Internal
+public final class DynamicWriter implements MessageBodyWriter<Object> {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/DynamicWriter.java]: Rename to `DynamicMessageBodyWriter`

File: test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy (hunk 16 above)
Comment (yawkat) [reply test-suite/src/test/groovy/io/micronaut/docs/http/client/proxy/ProxyRequestSpec.groovy:105]: TextPlainCodec used to just accept any object and call toString on it. TextPlainHandler only works with String, so it gives an error that there is no handler for text/plain for this type. i think it makes more sense.

File: http/src/main/java/io/micronaut/http/body/RawMessageBodyHandler.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Supported types of this raw body handler. Exact match is used for reading. For writing, the
+     * match is covariant. For example, if this returns {@code [String, CharSequence]}, then this
+     * raw handler will be used for reading types declared as exactly {@code String} or
+     * {@code CharSequence}, and will additionally be used for writing (but not reading) subtypes
+     * (e.g. {@code StringBuilder}).
+     *
+     * @return The supported types
+     */
+    Collection<? extends Class<?>> getTypes();
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/RawMessageBodyHandler.java:44]: nullability annotations for return type

File: http/src/main/java/io/micronaut/http/body/TextStreamBodyHandler.java [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * Handler for SSE events.
+ *
+ * @param <T> The type to write, may be {@link Event}
+ */
+@Internal
+@Singleton
+@Produces(MediaType.TEXT_EVENT_STREAM)
+@Consumes(MediaType.TEXT_EVENT_STREAM)
+public class TextStreamBodyHandler<T> implements MessageBodyWriter<T> {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/TextStreamBodyHandler.java]: since it is only a writer rename to `TextStreamBodyWriter` and make `final`

File: json-core/src/main/java/io/micronaut/json/body/JsonMessageHandler.java [hunk 29]
```diff
@@ -54,6 +54,10 @@ public JsonMessageHandler(JsonMapper jsonMapper) {
         this.jsonMapper = jsonMapper;
     }
 
+    public JsonMapper getJsonMapper() {
```
Comment (graemerocher) [inline json-core/src/main/java/io/micronaut/json/body/JsonMessageHandler.java:64]: javadoc / nullability annotations

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java (hunk 19 above)
Comment (yawkat) [reply http-client/src/main/java/io/micronaut/http/client/netty/FullNettyClientHttpResponse.java:164]: it's unused

File: http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java (hunk 24 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/body/ContextlessMessageBodyHandlerRegistry.java:39]: RawMessageBodyHandlerRegistry is package-private so moving ContextlessMessageBodyHandlerRegistry would need to make RawMessageBodyHandlerRegistry public

File: http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawMessageBodyHandler.java [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * @since 4.0.0
+ * @author Jonas Konrad
+ */
+@Internal
+@Singleton
+@Experimental
+@BootstrapContextCompatible
+@Bean(typed = RawMessageBodyHandler.class)
+public final class ByteBufRawMessageBodyHandler implements RawMessageBodyHandler<ByteBuf> {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawMessageBodyHandler.java:55]: does this need to be public?

File: http/src/main/java/io/micronaut/http/body/TextStreamBodyWriter.java [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * Handler for SSE events.
+ *
+ * @param <T> The type to write, may be {@link Event}
+ */
+@Internal
+@Singleton
+@Produces(MediaType.TEXT_EVENT_STREAM)
+@Consumes(MediaType.TEXT_EVENT_STREAM)
+public final class TextStreamBodyWriter<T> implements MessageBodyWriter<T> {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/body/TextStreamBodyWriter.java]: does this need to be public?

**Review by graemerocher — CHANGES_REQUESTED** [review]
Can you update `breaks.adoc` and document the breaking changes in this PR

File: http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawMessageBodyHandler.java (hunk 30 above)
Comment (yawkat) [reply http-netty/src/main/java/io/micronaut/http/netty/body/ByteBufRawMessageBodyHandler.java:55]: yes, it's used by the contextless http client

File: http/src/main/java/io/micronaut/http/body/DynamicWriter.java (hunk 5 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/body/DynamicWriter.java]: DynamicWriter is not shared. It used to be (the registry used to return it as fallback) but it didn't turn out to work so it's used directly now.

File: http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java (hunk 7 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java]: will use ObjectUtils. storing the hash shouldn't matter, the map caches it anyway.

File: http/src/main/java/io/micronaut/http/body/TextPlainHandler.java (hunk 8 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/body/TextPlainHandler.java:47]: don't think it matters since it's cached

**Review by graemerocher — APPROVED** [review]

**Review by dstepanov — APPROVED** [review]
